Initial persistence REST resource contribution - moved from HABmin

cdjackson opened the pull request:
Initial REST interface for persistence - moved over from HABmin bundle (and not tested yet - just moved, but it was heavily tested there :) ).

This provides two interfaces
`GET /rest/persistence/` which returns a list of persistence services
`GET /rest/persistence/{service}/{item}?{starttime=}&{endtime=}` which returns persistence data from the item for the specified service. 

As discussed on the forum, this is a straight copy of what I've been using with HABmin, so I welcome comments and am happy to change if needed.  If this looks like it's roughly what you're after, I'll tidy it up...

Signed-off-by: Chris Jackson <chris@cd-jackson.com>

mccare commented:
Get a cycle in the build path (see comment in code) and I got errors, that the ItemRegistry cannot be set on the PersistenceResource. Renaming the setItemUIRegistry to setItemRegistry got rid of this error. trying to hit /rest/persistence will still yield an empty array.

cdjackson commented:
I changed the provider when I moved it over, so it's possible that I messed something up, but it compiled ok for me. I will take another look...

I found with HABmin at least that it was quick to aggregate the data on a PC, and often not so quick to do it on the server, so I just send all the data between two times. One other point is that as each items data is sent in a separate request, the data processing can be done on the client while the server is reading the next parameters from the persistence store (of course this is only relevant when you're requesting multiple items).

mccare commented:
If you have a chance, it would be great if you could fire up OH2 and ESH in master, merge the changes and then just hit the GET /rest/persistence/. IMHO there should be some items there, right? In my test case this was empty.

cdjackson commented:
Sorry for the delay looking at this - I went on holiday for a few days :)

It seems to work ok for me - do you actually have any persistence services running in your development environment?

![image](https://cloud.githubusercontent.com/assets/1041475/11763178/a9d626e2-a0f8-11e5-9ba2-064465439c60.png)

maggu2810 commented:
Is this still something that is worked on and you would like more comments or something that could be seen as a mergable PR for review?

kaikreuzer commented:
I think it is waiting to be reviewed.

cdjackson commented:
From my perspective, it's waiting for review - I'm not working on it...

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/ItemHistoryListBean.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @XmlElement(name = "item")
+    public final List<ItemHistoryBean> entries = new ArrayList<ItemHistoryBean>();
```
I would prefer to place members above the constructor and other functions.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/ItemHistoryBean.java:
```diff
@@ -0,0 +1,111 @@
+    public ItemHistoryBean() {
+    };
+
+    public double addData(Long time, State state) {
```
Could you add JavaDoc, at least the return value needs an explanation.
For example: a numeric value that could represent the state. It could be used for e.g. min / max / avg calculation.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,289 @@
+@Api(value = PersistenceResource.PATH)
+public class PersistenceResource implements RESTResource {
+
+    private static final Logger logger = LoggerFactory.getLogger(PersistenceResource.class);
```
The logger member should not be static.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,289 @@
+        persistenceServices.remove(service.getName());
+    }
+
+    static public Map<String, PersistenceService> getPersistenceServices() {
```
The function could be removed.
* no static access to private member is used here
* internally we could use the member variable directly.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,289 @@
+        return Response.ok(responseObject).build();
+    }
+
+    private Date convertTime(String sTime) {
```
Why not using the DateTimeType, so the user of the REST API can use timezone and we can remove duplicate code:

```java
DateTimeType dateTime = new DateTimeType(sTime);
return dateTime.getCalendar().getTime();
```

maggu2810 commented:
IMHO we should remove the most of the comment code.

cdjackson commented:
> On 14 Jan 2016, at 14:28, Markus Rathgeb <notifications@github.com> wrote:
> 
> IMHO we should remove the most of the comment code.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/eclipse/smarthome/pull/549#issuecomment-171657516>.
> 
Agreed - most of this I simply commented out when I copied the source from HABmin. As per my original comment I wanted to make sure this was of interest before I tidied :)

cdjackson commented:
Updates complete...

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/META-INF/MANIFEST.MF:
```diff
@@ -6,8 +6,20 @@ Bundle-Version: 0.8.0.qualifier
+ org.eclipse.smarthome.core.thing,
+ org.eclipse.smarthome.core.types,
+ org.eclipse.smarthome.io.rest,
+ org.eclipse.smarthome.ui.items,
```
This dependency from a core bundle to a ui bundle is ugly and it does not seem to be at all required...?

cdjackson replied:
I think this is not used now - it's a hangover from Habmin 1 so will remove. 

Sent from my iPhone

> On 20 Jan 2016, at 13:52, Kai Kreuzer <notifications@github.com> wrote:
> 
> In bundles/core/org.eclipse.smarthome.core.persistence/META-INF/MANIFEST.MF:
> 
> >   org.eclipse.smarthome.core.persistence,
> > - org.eclipse.smarthome.core.types
> > + org.eclipse.smarthome.core.thing,
> > + org.eclipse.smarthome.core.types,
> > + org.eclipse.smarthome.io.rest,
> > + org.eclipse.smarthome.ui.items,
> This dependency from a core bundle to a ui bundle is ugly and it does not seem to be at all required...?
> 
> —
> Reply to this email directly or view it on GitHub.
>

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/META-INF/MANIFEST.MF:
```diff
@@ -6,8 +6,20 @@ Bundle-Version: 0.8.0.qualifier
 Bundle-License: http://www.eclipse.org/legal/epl-v10.html
 Bundle-SymbolicName: org.eclipse.smarthome.core.persistence
-Import-Package: org.eclipse.smarthome.core.items,
+Import-Package: io.swagger.annotations,
```
please mark this as an optional import, so that it only has to be present at compile time, but not runtime.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/META-INF/MANIFEST.MF:
```diff
@@ -6,8 +6,20 @@ Bundle-Version: 0.8.0.qualifier
+Import-Package: io.swagger.annotations,
+ javax.servlet,
+ javax.servlet.http,
+ javax.ws.rs,
```
please make all rest related imports optional, so that the bundle is also usable without REST support and will then simply not activate the rest resource component.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/META-INF/MANIFEST.MF:
```diff
@@ -6,8 +6,20 @@ Bundle-Version: 0.8.0.qualifier
+ org.eclipse.smarthome.io.rest,
+ org.eclipse.smarthome.ui.items,
+ org.slf4j
+Require-Bundle: javax.servlet
```
never ever do a require-bundle!

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/ItemHistoryBean.java:
```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) 2010-2013, openHAB.org and others.
```
please update the license headers with `mvn license:format`.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/ItemHistoryBean.java:
```diff
@@ -0,0 +1,88 @@
+import org.eclipse.smarthome.core.types.State;
+
+/**
+ * This is a java bean that is used with JAXB to serialize items to JSON.
```
Nope, no JAXB involved - GSON does the marshalling.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/ItemHistoryListBean.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.Collection;
+import java.util.List;
+
+import javax.xml.bind.annotation.XmlElement;
```
please remove all the javax.xml.bind stuff here.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/ItemHistoryListBean.java:
```diff
@@ -0,0 +1,35 @@
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * This is a java bean that is used with JAXB to serialize item lists.
```
no JAXB

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/ServiceBean.java:
```diff
@@ -0,0 +1,19 @@
+package org.eclipse.smarthome.core.persistence.rest;
+
+/**
+ * This is a java bean that is used with JAXB to serialize items to JSON.
```
no JAXB

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/ServiceBean.java:
```diff
@@ -0,0 +1,19 @@
+ *
+ */
+public class ServiceBean {
+    public String name;
```
We should probably work on the persistence services in future to also have some more meta-data like a localizable label and description for the UI...

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    @GET
+    @Path("/{servicename: [a-zA-Z_0-9]*}/{itemname: [a-zA-Z_0-9]*}")
```
I would suggest to have servicename as an optional queryparam only.

kaikreuzer replied:
In order to retrieve the configured defaultService, we should probably move the configuration from PersistenceExtensions.defaultService to some new exported class in core.persistence, so that this info can be used from different places.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,261 @@
+    @Path("/{servicename: [a-zA-Z_0-9]*}/{itemname: [a-zA-Z_0-9]*}")
+    @Produces({ MediaType.APPLICATION_JSON })
+    @ApiOperation(value = "Gets item persistence data from the persistence service.", response = ItemHistoryBean.class)
+    @ApiResponses(value = @ApiResponse(code = 200, message = "OK") )
```
hm, I assume besides HTTP 200 there can be also many error situations that should be listed here? (like item unknown -> 404, endTime<starttime -> 400, etc.)

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,261 @@
+    @ApiOperation(value = "Gets item persistence data from the persistence service.", response = ItemHistoryBean.class)
+    @ApiResponses(value = @ApiResponse(code = 200, message = "OK") )
+    public Response httpGetPersistenceItemData(@Context HttpHeaders headers,
+            @PathParam("servicename") String serviceName, @PathParam("itemname") String itemName,
```
would be nice to have some swagger metadata (description, required, etc.) for the params as well.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,261 @@
+
+        if (service == null) {
+            logger.debug("Persistence service not found '{}'.", serviceName);
+            throw new WebApplicationException(404);
```
Please use JSONResponse here, so that you can also add an error message in it.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,261 @@
+
+        if (!(service instanceof QueryablePersistenceService)) {
+            logger.debug("Persistence service not queryable '{}'.", serviceName);
+            throw new WebApplicationException(404);
```
Please use JSONResponse here, so that you can also add an error message in it.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/ItemHistoryBean.java:
```diff
@@ -0,0 +1,88 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.persistence.rest;
```
as the package is not exported, you should move it to persistence.internal.rest

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,261 @@
+        if (result != null && result.iterator().hasNext()) {
+            HistoricItem historicItem = result.iterator().next();
+
+            double value = bean.addData(dateTimeBegin.getTime(), historicItem.getState());
```
Why do you use the unprecise double here instead of continuing to work with BigDecimal, which is used internally in the state?

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,261 @@
+        State state = null;
+
+        Long quantity = 0l;
+        double average = 0;
```
The average value should be treated with care by the client - it is the average over the datapoints, no the average value over time (think of 10 datapoints within the first second and one further an hour later).

cdjackson replied:
I guess that's also the same for min/max - depending on the persistence service being used (eg rrd doesn't log exact data).
I'm wondering if I should remove these statistical parts of the response - what do you think? HABmin doesn't use it so I don't care and it might be safer to return the data and let the client do whatever it likes to calculate/present this if it wants?

kaikreuzer replied:
I'm ok with removing them, yes.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,261 @@
+        }
+
+        long timerStop = System.currentTimeMillis();
+        logger.debug("CHART: returned {} rows in {}ms", bean.datapoints, timerStop - timerStart);
```
Not "CHART" - this is only a time series rest resource :-)

kaikreuzer commented:
Thanks Chris, and sorry for taking so long for the review!

Besides my review comments, I have a general question:
The current implementation returns all available datapoints within a given period - which is fine, because this is what the QueryablePersistenceServices provide.

I wonder if for charting purposes it might be useful to have a possibility to
- retrieve equidistant datapoints, e.g. one for every minute
- optionally retrieve viewer results, e.g. when rendering a chart for the past year (which would be solved through equidistant datapoints as well)

This is likely not too easy to implement, since you need to have different strategies on how to condense the datapoints - I know from rrd that there is a whole science behind that... But I'd like to raise this discussion, especially since you have already done JS charts in HABmin, so you have experience about how usable the plain list of datapoints is.

cdjackson commented:
Hi Kai,
Regarding your point about subsampling (eg one per minute) - I don't think that this is something that should be in the REST implementation...

My thoughts - it's a wider persistence issue -:

The REST interface should return what the persistence service provides. We could subsample, but when I first implemented this interface a couple of years ago, I tried this and found it was much slower than just returning all the data from the server, and then letting the client do the subsampling if needed. Since this is a REST interface, it's normally Java, or javascript, or C or... that interfaces to it, so this is no problem and it's faster (in my experience - especially with something like a PI!). In HABmin, I find this works quite fast - I also think that the client should be 'smarter' so that it updates data from the event source rather than requesting an update of the whole chart every 30 seconds (for example). Expanding on this, if you zoom out, then the client should only request the new data - not the whole period. I think to keep things responsive for the user, both the client and server side need to take some "responsibility".

If we wanted to have (for example) data every minute, then I think we should do this in the persistence manager (or somewhere like this) so that it's consistent across all services. I would then prefer to store multiple sets of data into the persistence store (as rrd does already) rather than having to subsample, or process this data in real-time which would likely seriously impact performance. I know this means more memory, but (generally!) this shouldn't be an issue these days - memory is pretty cheap, but people tend to be impatient and want to see a graph quickly :)

Moving on from the above, I would like to see an extended REST store() method that also takes a time. I've got a half written email suggesting a store(HistoricState) method so that we can pass in time. It would then be possible to implement other persistence solutions that provided a common interface for storing (for example) daily, hourly (etc) data - this is something that is quite often asked for - IMHO it needs to be done outside of the REST interface so that it is responsive.

Anyway.... My suggestion would be to keep the persistence REST interface for reading the data  simple, and then look to add features for subsampling (etc) into a separate bundle (ie persistencemanager, or in that area). The REST interface can then be extended to allow the user to specify what data 'type' they want (ie hourly, daily, average etc). 

I've probably suggested a different avenue than you were thinking though :)

kaikreuzer commented:
Thanks for your input, Chris.
Ok, this is fine for me - Someone who uses his SQL database to store 10 temperature values per second and then wants to render a chart for the last decade will notice that this isn't a good idea and that a persistence service like rrd might be a better choice for such use cases.

Regarding store(HistoricState) - let's do that discussion outside of this PR, it might otherwise not get merged anytime soon ;-)

cdjackson commented:
> On 20 Jan 2016, at 16:26, Kai Kreuzer <notifications@github.com> wrote:
> 
> Regarding store(HistoricState) - let's do that discussion outside of this PR, it might otherwise not get merged anytime soon ;-)

Agreed - I didn’t really want to discuss it here, but wanted to raise the idea of doing something outside of the rest interface… I’ll look at addressing that on the ESH forum soon...

cdjackson commented:
Not sure why this has picked up some other files, so it probably needs a rebase!

Otherwise this should cover most of your comments - the outstanding one is the default persistence service (and I need to test it still works after the changes!).

maggu2810 commented:
Should the caller be able to set a "max data points that are returned" limit?
I have needed this on other solutions. If you do not know how many data points has been persistent, you could force that the reply does not contain more then x values. The reply contains a indication if all values are returned or if there are remaining ones. The caller could then decide if he takes the timestamp of the last value and asks again (this timestamp is the new start time and the end time keeps the same) etc.

cdjackson commented:
This should already be possible by using the paging - set the page size to the number of entries you want returned, and ask for the first page?

I’ve never used paging, so can’t confirm it works, but in theory, this should do what you want.

maggu2810 commented:
What I named "the caller" is the user of the REST API. Very small devices perhaps want a much lower limit of returned data points then my high power machine.

cdjackson commented:
I understand what you mean by caller - but not what your comment means? Do you not think my suggestion works? (as I said, I've not actually checked)

cdjackson commented:
Actually, paging doesn't appear to be supported other than through the REST API - I'll update this - I think this does what you want?

maggu2810 commented:
Perhaps I read to much code this week. Have missed the pages (does not know why) on reading. All fine. :wink:

cdjackson commented:
:)

cdjackson commented:
@kaikreuzer how do you want the default persistence service handled? I don't mind changing this if you like, although if someone else more familiar with that part of the code can handle it it might be better...

kaikreuzer commented:
Feel free to leave that out of this PR and I will see if I can come up with a good solution.

cdjackson commented:
Ok, in that case I think it's basically done. I'll have a look through it over the next few days to make sure it's ok, and given there's no default provider I'll make sure this is covered for now.

When I'm done I'll squash the commits...

cdjackson commented:
I think this is good to merge - I've now tested this against an updated version of HABmin.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/internal/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,237 @@
+            dateTimeBegin = new Date(dateTimeEnd.getTime() - MILLISECONDS_PER_DAY);
+        }
+
+        // Default to 1 days data if the times are the same
```
Default the start to one day in front of the end if the times are the same or the start points to a time after the end.

cdjackson replied:
Do you mean you want me to change the comment - doesn’t it say the same thing in a different way?

maggu2810 replied:
Yes, I would change the comment that it reflects the code. Tthe current comment misses that also start > end is handled and I would add how it is handled (not the end is changed, but the begin).
My comment above should be an example one.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/internal/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,237 @@
+
+        // If serviceName is null, then use the default service
+        PersistenceService service = null;
+        if (serviceName == null) {
```
> Feel free to leave that out of this PR and I will see if I can come up with a good solution.

Perhaps we should move the comment (in the code) to the branch and change the comment that it does not state that a default is used ATM.

```diff
+        PersistenceService service = null;
+        if (serviceName == null) {
+            // Add handing (e.g. use default persistence service) here some time.
 +        } else {
+            service = persistenceServices.get(serviceName);
+        }
```

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/internal/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,237 @@
+        Long quantity = 0l;
+
+        ItemHistoryBean bean = null;
+        bean = new ItemHistoryBean();
```
Why set bean to null in the one line and then using new ItemHistoryBean in the next line?

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/internal/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,237 @@
+            dateTimeBegin = new Date(dateTimeEnd.getTime() - MILLISECONDS_PER_DAY);
+        }
+
+        FilterCriteria filter = new FilterCriteria();
```
You can drop this new (or the below) as this object is thrown away without using it.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/internal/rest/PersistenceResource.java:
```diff
@@ -0,0 +1,237 @@
+    /**
+     * Gets a list of persistence services currently configured in the system
+     *
+     * @return
```
Missing JavaDoc what is returned ;)

maggu2810 commented:
LGTM, just a few minor remarks.

cdjackson commented:
Updated...
